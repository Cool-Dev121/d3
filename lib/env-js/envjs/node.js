require('./platform/node');
Envjs.eventLoop = function() {}; // disabled for clean shutdown

require('./window');
new Window(__this__); // initialize a blank window
