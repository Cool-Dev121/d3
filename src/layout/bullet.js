// ranges (bad, satisfactory, good)
// measures (actual, forecast)
// markers (previous, goal)

/*
 * Chart design based on the recommendations of Stephen Few. Implementation
 * based on the work of Clint Ivy, Jamie Love, and Jason Davies.
 * http://projects.instantcognition.com/protovis/bulletchart/
 */

/**
 * Constructs a new, empty bullet layout.
 */
d3.layout.bullet = function() {
  var orient = 'left',
      ranges = [],
      markers = [],
      measures = [],
      horizontal,
      maximum = null,
      rangeColor = d3.scale.linear(),
      measureColor = d3.scale.linear(),
      scale = d3.scale.linear();

  var bullet = function() {
    ranges.sort(function(a, b) { return b - a });
    measures.sort(function(a, b) { return b - a });
    this.selectAll('rect.range')
      .data(ranges)
        .enter().append('svg:rect')
        .attr('class', 'range')
        .attr('width', function(d) { return scale(d) + '%' })
        .attr('height', '100%')
        .attr('style', function(d, i) { return 'fill:' + rangeColor(i) });
    this.selectAll('rect.measure')
      .data(measures)
        .enter().append('svg:rect')
        .attr('class', 'measure')
        .attr('width', function(d) { return scale(d) + '%' })
        .attr('height', '34%')
        .attr('y', '33%')
        .attr('fill', function(d, i) { return measureColor(i) });
    this.selectAll('line.marker')
      .data(markers)
        .enter().append('svg:line')
        .attr('class', 'marker')
        .attr('x1', function(d) { return scale(d) + '%' })
        .attr('x2', function(d) { return scale(d) + '%' })
        .attr('y1', '15%')
        .attr('y2', '85%')
        .attr('stroke', '#000')
        .attr('stroke-width', '2px')
  }

  /** Cache layout state to optimize properties. */
  var buildCache = function() {
    horizontal = /^left|right$/.test(orient);
    rangeColor.domain([0, Math.max(1, ranges.length - 1)])
        .range(["#eee", "#bbb"]);
    measureColor.domain([0, Math.max(1, measures.length - 1)])
        .range(["lightsteelblue", "steelblue"]);
    maximum = maximum || d3.max([].concat(ranges, markers, measures));
    scale.domain([0, maximum]).range([0, 100]);
  };

  // left, right, top, bottom
  bullet.orient = function(x) {
    if (!arguments.length) return orient;
    orient = x;
    buildCache();
    return bullet;
  };

  bullet.ranges = function(x) {
    if (!arguments.length) return ranges;
    ranges = x;
    buildCache();
    return bullet;
  };

  bullet.markers = function(x) {
    if (!arguments.length) return markers;
    markers = x;
    buildCache();
    return bullet;
  };

  bullet.measures = function(x) {
    if (!arguments.length) return measures;
    measures = x;
    buildCache();
    return bullet;
  };

  bullet.maximum = function(x) {
    if (!arguments.length) return maximum;
    maximum = x;
    buildCache();
    return bullet;
  };

  buildCache();
  return bullet;
};
