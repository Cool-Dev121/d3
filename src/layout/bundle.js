// Implements hierarchical edge bundling using Holten's algorithm. For each
// edge, an open uniform b-spline is computed that travels through the tree, up
// the parent hierarchy to the least common ancestor, and then back down to the
// destination node.
d3.layout.bundle = function() {
  return function(links) {
    var splines = [],
        i = -1,
        n = links.length;
    while (++i < n) splines.push(d3_layout_bundleSpline(links[i]));
    return splines;
  };
};

function d3_layout_bundleSpline(link) {
  var start = link.source,
      end = link.target,
      lca = d3_layout_bundleLeastCommonAncestor(start, end),
      points = [start];
  while (start !== lca) {
    start = start.parent;
    points.push(start);
  }
  var k = points.length;
  while (end !== lca) {
    points.splice(k, 0, end);
    end = end.parent;
  }
  return points;
}

function d3_layout_bundleAncestors(node) {
  var ancestors = [],
      parent = node.parent;
  while (parent != null) {
    ancestors.push(node);
    node = parent;
    parent = parent.parent;
  }
  ancestors.push(node);
  return ancestors;
}

function d3_layout_bundleLeastCommonAncestor(a, b) {
  if (a === b) return a;
  var aNodes = d3_layout_bundleAncestors(a),
      bNodes = d3_layout_bundleAncestors(b),
      aNode = aNodes.pop(),
      bNode = bNodes.pop(),
      sharedNode = null;
  while (aNode === bNode) {
    sharedNode = aNode;
    aNode = aNodes.pop();
    bNode = bNodes.pop();
  }
  return sharedNode;
}
