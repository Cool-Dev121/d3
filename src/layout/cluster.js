// Implements a hierarchical layout using the cluster (or dendogram) algorithm.
d3.layout.cluster = function() {
  var hierarchy = d3.layout.hierarchy(),
      group = 0;

  function cluster(d, i) {
    var nodes = hierarchy.call(this, d, i),
        root = nodes[0],
        leafCount = 0,
        leafIndex = .5 - group / 2;

    /* Count the leaf nodes and compute the depth of descendants. */
    var p = undefined;
    d3_layout_clusterVisitAfter(root, function(n) {
      if (n.children) {
        n.depth = 1 + d3.max(n.children, function(n) { return n.depth; });
      } else {
        if (group && (p != n.parent)) {
          p = n.parent;
          leafCount += group;
        }
        leafCount++;
        n.depth = 0;
      }
    });
    var breadth = 1 / leafCount;
    var depth = 1 / root.depth;

    /* Compute the unit breadth and depth of each node. */
    var p = undefined;
    d3_layout_clusterVisitAfter(root, function(n) {
      if (n.children) {
        n.breadth = d3_layout_clusterMean(n.children, function(n) { return n.breadth; });
      } else {
        if (group && (p != n.parent)) {
          p = n.parent;
          leafIndex += group;
        }
        n.breadth = breadth * leafIndex++;
      }
      n.depth = 1 - n.depth * depth;
    });

    /* Compute breadth and depth ranges for space-filling layouts. */
    d3_layout_clusterVisitAfter(root, function(n) {
      n.minBreadth = n.children
          ? n.children[0].minBreadth
          : (n.breadth - breadth / 2);
      n.maxBreadth = n.children
          ? n.children[n.children.length - 1].maxBreadth
          : (n.breadth + breadth / 2);
    });
    d3_layout_clusterVisitBefore(root, function(n) {
      n.minDepth = n.parent
          ? n.parent.maxDepth
          : 0;
      n.maxDepth = n.parent
          ? (n.depth + root.depth)
          : (n.minDepth + 2 * root.depth);
    });
    root.minDepth = -depth;

    return nodes;
  }

  cluster.sort = d3.rebind(cluster, hierarchy.sort);
  cluster.children = d3.rebind(cluster, hierarchy.children);
  cluster.value = d3.rebind(cluster, hierarchy.value);

  cluster.group = function(x) {
    if (!arguments.length) return group;
    group = x;
    return cluster;
  };

  return cluster;
};

d3_layout_clusterVisitAfter = d3_layout_treeVisitAfter;

function d3_layout_clusterVisitBefore(node, callback) {
  function visit(node, previousSibling) {
    callback(node, previousSibling);
    var children = node.children;
    if (children) {
      var child,
          previousChild = null,
          i = -1,
          n = children.length;
      while (++i < n) {
        child = children[i];
        visit(child, previousChild);
        previousChild = child;
      }
    }
  }
  visit(node, null);
}

function d3_layout_clusterSum(array, f) {
  var o = {};
  return array.reduce(f
      ? function(p, d, i) { o.index = i; return p + f.call(o, d); }
      : function(p, d) { return p + d; }, 0);
}

function d3_layout_clusterMean(array, f) {
  return d3_layout_clusterSum(array, f) / array.length;
}
