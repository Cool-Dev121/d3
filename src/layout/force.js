// A rudimentary force layout using Gauss-Seidel.
d3.layout.force = function() {
  var force = {},
      event = d3.dispatch("tick"),
      size = [1, 1],
      alpha,
      drag = .9,
      distance = 30,
      charge = -60,
      gravity = .02,
      theta = .8,
      interval,
      nodes,
      links,
      distances;

  function accumulate(n) {
    var cx = 0,
        cy = 0;
    n.count = 0;
    if (!n.leaf) {
      n.nodes.forEach(function(c) {
        accumulate(c);
        n.count += c.count;
        cx += c.count * c.cx;
        cy += c.count * c.cy;
      });
    }
    if (n.point) {
      n.count++;
      cx += n.point.x;
      cy += n.point.y;
    }
    n.cx = cx / n.count;
    n.cy = cy / n.count;
  }

  function repulse(p) {
    return function(n, x1, y1, x2, y2) {
      if (n.point != p) {
        var dx = (n.cx - p.x) || Math.random(),
            dy = (n.cy - p.y) || Math.random(),
            dn = 1 / Math.sqrt(dx * dx + dy * dy);

        /* Barnes-Hut criterion. */
        if ((x2 - x1) * dn < theta) {
          var k = alpha * charge * n.count * dn * dn;
          p.fx += dx * k;
          p.fy += dy * k;
          return true;
        }

        if (n.point) {
          var k = alpha * charge * dn * dn;
          p.fx += dx * k;
          p.fy += dy * k;
        }
      }
    };
  }

  function tick() {
    var n = nodes.length,
        m = links.length,
        q = d3.geom.quadtree(nodes),
        i, // current index
        o, // current object
        s, // current source
        t, // current target
        l, // current distance
        x, // x-distance
        y; // y-distance

    // reset forces
    i = -1; while (++i < n) {
      (o = nodes[i]).fx = o.fy = 0;
    }

    // gauss-seidel relaxation for links
    for (i = 0; i < m; ++i) {
      o = links[i];
      s = o.source;
      t = o.target;
      x = t.x - s.x;
      y = t.y - s.y;
      if (l = Math.sqrt(x * x + y * y)) {
        l = alpha * (l - distance) / l;
        x *= l;
        y *= l;
        t.x -= x;
        t.y -= y;
        s.x += x;
        s.y += y;
      }
    }

    // compute quadtree center of mass
    accumulate(q);

    // apply gravity forces
    x = size[0] / 2;
    y = size[1] / 2;
    i = -1; while (++i < n) {
      o = nodes[i];
      s = x - o.x;
      t = y - o.y;
      l = alpha * gravity / n * Math.sqrt(s * s + t * t);
      s *= l;
      t *= l;
      o.fx += s;
      o.fy += t;
    }

    // apply charge forces
    i = -1; while (++i < n) {
      q.visit(repulse(nodes[i]));
    }

    // position verlet integration
    i = -1; while (++i < n) {
      o = nodes[i];
      if (o.fixed) {
        o.x = o.px;
        o.y = o.py;
      } else {
        x = o.px - (o.px = o.x);
        y = o.py - (o.py = o.y);
        o.x += o.fx - x * drag;
        o.y += o.fy - y * drag;
      }
    }

    event.tick.dispatch({type: "tick"});

    // simulated annealing, basically
    return (alpha *= .99) < .005;
  }

  force.on = function(type, listener) {
    event[type].add(listener);
    return force;
  };

  force.nodes = function(x) {
    if (!arguments.length) return nodes;
    nodes = x;
    return force;
  };

  force.links = function(x) {
    if (!arguments.length) return links;
    links = x;
    return force;
  };

  force.size = function(x) {
    if (!arguments.length) return size;
    size = x;
    return force;
  };

  force.distance = function(d) {
    if (!arguments.length) return distance;
    distance = d;
    return force;
  };

  force.start = function() {
    var i,
        n = nodes.length,
        m = links.length,
        w = size[0],
        h = size[1],
        o;

    // TODO initialize positions of new nodes using constraints (links)
    for (i = 0; i < n; ++i) {
      o = nodes[i];
      if (isNaN(o.x)) o.x = Math.random() * w;
      if (isNaN(o.y)) o.y = Math.random() * h;
      if (isNaN(o.px)) o.px = o.x;
      if (isNaN(o.py)) o.py = o.y;
    }

    for (i = 0; i < m; ++i) {
      o = links[i];
      if (typeof o.source == "number") o.source = nodes[o.source];
      if (typeof o.target == "number") o.target = nodes[o.target];
    }

    alpha = .1;
    d3.timer(tick);
    return force;
  };

  force.resume = force.start;

  force.stop = function() {
    alpha = 0;
    return force;
  };

  // use `node.call(force.drag)` to make nodes draggable
  force.drag = function() {

    this
      .on("mouseover", d3_layout_forceDragOver)
      .on("mouseout", d3_layout_forceDragOut)
      .on("mousedown", d3_layout_forceDragDown);

    d3.select(window)
      .on("mousemove", dragmove)
      .on("mouseup", dragup);

    return force;
  };

  function dragmove() {
    if (!d3_layout_forceDragNode) return;
    var m = d3.svg.mouse(d3_layout_forceDragElement);
    d3_layout_forceDragNode.px = m[0];
    d3_layout_forceDragNode.py = m[1];
    force.start(); // restart annealing
  }

  function dragup() {
    if (!d3_layout_forceDragNode) return;
    dragmove();
    d3_layout_forceDragNode.fixed = false;
    d3_layout_forceDragNode = d3_layout_forceDragElement = null;
  }

  return force;
};

var d3_layout_forceDragNode,
    d3_layout_forceDragElement;

function d3_layout_forceDragOver(d) {
  d.fixed = true;
}

function d3_layout_forceDragOut(d) {
  if (d !== d3_layout_forceDragNode) {
    d.fixed = false;
  }
}

function d3_layout_forceDragDown(d) {
  (d3_layout_forceDragNode = d).fixed = true;
  d3_layout_forceDragElement = this;
  d3.event.preventDefault();
}
