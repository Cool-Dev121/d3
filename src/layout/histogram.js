d3.layout.histogram = function() {
  var frequency = true,
      value = Number,
      ticksFunction = d3_layout_histogramTicks;

  function histogram(data, i) {
    var x = data.map(value), bins = [];

    // Initialize default ticks.
    var ticks = ticksFunction.call(this, data, i);

    // Initialize the bins.
    for (var i = 0; i < ticks.length - 1; i++) {
      var bin = bins[i] = [];
      bin.x = ticks[i];
      bin.dx = ticks[i + 1] - ticks[i];
      bin.y = 0;
    }

    // Count the number of samples per bin.
    for (var i = 0; i < x.length; i++) {
      var j = d3_layout_histogramSearchIndex(ticks, x[i]) - 1,
          bin = bins[Math.max(0, Math.min(bins.length - 1, j))];
      bin.y++;
      bin.push(data[i]);
    }

    // Convert frequencies to probabilities.
    if (!frequency) for (var i = 0; i < bins.length; i++) {
      bins[i].y /= x.length;
    }

    return bins;
  }

  histogram.frequency = function(x) {
    if (!arguments.length) return frequency;
    frequency = Boolean(x);
    return histogram;
  };

  histogram.ticks = function(x) {
    if (!arguments.length) return ticksFunction;
    ticksFunction = d3.functor(x);
    return histogram;
  };

  histogram.value = function(x) {
    if (!arguments.length) return value;
    value = x;
    return histogram;
  };

  return histogram;
};

// Performs a binary search on a sorted array.
// Returns the index of the value if found, otherwise -(insertion point) - 1.
// The insertion point is the index at which value should be inserted into the
// array for the array to remain sorted.
function d3_layout_histogramSearch(array, value) {
  var low = 0, high = array.length - 1;
  while (low <= high) {
    var mid = (low + high) >> 1, midValue = array[mid];
    if (midValue < value) low = mid + 1;
    else if (midValue > value) high = mid - 1;
    else return mid;
  }
  return -low - 1;
}

function d3_layout_histogramSearchIndex(array, value) {
  var i = d3_layout_histogramSearch(array, value);
  return (i < 0) ? (-i - 1) : i;
}

function d3_layout_histogramTicks(x) {
  return d3.scale.linear().domain(x).ticks(10);
}
