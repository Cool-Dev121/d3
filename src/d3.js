import "start";
import "core/core";

// Compatibility
import "compat/date";
import "compat/style";

// Arrays
import "core/ascending";
import "core/descending";
import "core/min";
import "core/max";
import "core/extent";
import "core/sum";
import "core/mean";
import "core/median";
import "core/quantile";
import "core/bisect";
import "core/shuffle";
import "core/permute";
import "core/zip";
import "core/transpose";
import "core/keys";
import "core/values";
import "core/entries";
import "core/merge";
import "core/range";
import "core/nest";
import "core/map";
import "core/set";

// Math
import "core/random";

// String Formatting
import "core/requote";
import "core/round";

// Loading External Resources
import "core/xhr";
import "core/text";
import "core/json";
import "core/html";
import "core/xml";

// Namespaces
import "core/ns";

// Internals
import "core/functor";
import "core/rebind";
import "core/dispatch";

///////////




import "core/format";
import "core/formatPrefix";
import "core/ease";
import "core/event";
import "core/transform";
import "core/interpolate";
import "core/uninterpolate";
import "core/color";
import "core/rgb";
import "core/hsl";
import "core/hcl";
import "core/lab";
import "core/xyz";
import "core/selection";
import "core/selection-select";
import "core/selection-selectAll";
import "core/selection-attr";
import "core/selection-classed";
import "core/selection-style";
import "core/selection-property";
import "core/selection-text";
import "core/selection-html";
import "core/selection-append";
import "core/selection-insert";
import "core/selection-remove";
import "core/selection-data";
import "core/selection-datum";
import "core/selection-filter";
import "core/selection-order";
import "core/selection-sort";
import "core/selection-on";
import "core/selection-each";
import "core/selection-call";
import "core/selection-empty";
import "core/selection-node";
import "core/selection-transition";
import "core/selection-root";
import "core/selection-enter";
import "core/selection-enter-select";
import "core/transition";
import "core/transition-select";
import "core/transition-selectAll";
import "core/transition-filter";
import "core/transition-attr";
import "core/transition-style";
import "core/transition-text";
import "core/transition-remove";
import "core/transition-ease";
import "core/transition-delay";
import "core/transition-duration";
import "core/transition-each";
import "core/transition-transition";
import "core/transition-tween";
import "core/timer";
import "core/mouse";
import "core/touches";
import "core/noop";

import "scale/scale";
import "scale/nice";
import "scale/linear";
import "scale/bilinear";
import "scale/polylinear";
import "scale/log";
import "scale/pow";
import "scale/sqrt";
import "scale/ordinal";
import "scale/category";
import "scale/quantile";
import "scale/quantize";
import "scale/threshold";
import "scale/identity";

import "svg/svg";
import "svg/arc";
import "svg/line";
import "svg/line-radial";
import "svg/area";
import "svg/area-radial";
import "svg/chord";
import "svg/diagonal";
import "svg/diagonal-radial";
import "svg/symbol";
import "svg/axis";
import "svg/brush";

import "behavior/behavior";
import "behavior/drag";
import "behavior/zoom";

import "layout/layout";
import "layout/bundle";
import "layout/chord";
import "layout/force";
import "layout/partition";
import "layout/pie";
import "layout/stack";
import "layout/histogram";
import "layout/hierarchy";
import "layout/pack";
import "layout/cluster";
import "layout/tree";
import "layout/treemap";
import "layout/voronoi";

import "geo/geo";
import "geo/area";
import "geo/bounds";
import "geo/cartesian";
import "geo/centroid";
import "geo/circle";
import "geo/clip";
import "geo/clip-antimeridian";
import "geo/clip-circle";
import "geo/clip-polygon";
import "geo/clip-view";
import "geo/compose";
import "geo/distance";
import "geo/graticule";
import "geo/haversin";
import "geo/greatArc";
import "geo/interpolate";
import "geo/length";
import "geo/path-area";
import "geo/path-buffer";
import "geo/path-centroid";
import "geo/path-context";
import "geo/path";
import "geo/projection";
import "geo/resample";
import "geo/rotation";
import "geo/spherical";
import "geo/stream";
import "geo/albers";
import "geo/albers-usa";
import "geo/azimuthal";
import "geo/azimuthal-equal-area";
import "geo/azimuthal-equidistant";
import "geo/conic";
import "geo/conic-conformal";
import "geo/conic-equal-area";
import "geo/conic-equidistant";
import "geo/equirectangular";
import "geo/gnomonic";
import "geo/mercator";
import "geo/orthographic";
import "geo/stereographic";
import "geo/transverse-mercator";

import "dsv/dsv";
import "dsv/csv";
import "dsv/tsv";

import "time/time";
import "time/format-localized";
import "time/format";
import "time/format-utc";
import "time/format-iso";
import "time/interval";
import "time/second";
import "time/minute";
import "time/hour";
import "time/day";
import "time/week";
import "time/month";
import "time/year";
import "time/scale";
import "time/scale-utc";

import "geom/geom";
import "geom/hull";
import "geom/polygon";
import "geom/voronoi";
import "geom/delaunay";
import "geom/quadtree";

import "end";
