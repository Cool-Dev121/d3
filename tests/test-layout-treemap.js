require("./../lib/env-js/envjs/node");
require("./../d3");
require("./../d3.layout");

var treemap = d3.layout.treemap();

console.log("zero, negative and NaN values:");
console.log("{value: 0}   ", treemap({value: 0})[0].area);
console.log("{value: -1}  ", treemap({value: -1})[0].area);
console.log("{value: NaN} ", treemap({value: NaN})[0].area);
console.log("");

treemap.size([0, 0]);
console.log("size([0, 0]):");
console.log("{value: 0}   ", treemap({value: 0})[0].area);
console.log("{value: -1}  ", treemap({value: -1})[0].area);
console.log("{value: NaN} ", treemap({value: NaN})[0].area);
console.log("{value: 1}   ", treemap({value: 1})[0].area);
console.log("");

treemap.size([-1, -1]);
console.log("size([-1, -1]):");
console.log("{value: 0}   ", treemap({value: 0})[0].area);
console.log("{value: -1}  ", treemap({value: -1})[0].area);
console.log("{value: NaN} ", treemap({value: NaN})[0].area);
console.log("{value: 1}   ", treemap({value: 1})[0].area);
console.log("");

treemap.size([-1, 1]);
console.log("size([-1, 1]):");
console.log("{value: 0}   ", treemap({value: 0})[0].area);
console.log("{value: -1}  ", treemap({value: -1})[0].area);
console.log("{value: NaN} ", treemap({value: NaN})[0].area);
console.log("{value: 1}   ", treemap({value: 1})[0].area);
console.log("");

treemap.size([NaN, NaN]);
console.log("size([NaN, NaN]):");
console.log("{value: 0}   ", treemap({value: 0})[0].area);
console.log("{value: -1}  ", treemap({value: -1})[0].area);
console.log("{value: NaN} ", treemap({value: NaN})[0].area);
console.log("{value: 1}   ", treemap({value: 1})[0].area);
console.log("");
