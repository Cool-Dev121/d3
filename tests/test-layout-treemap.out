zero, negative and NaN values:
{value: 0}    0
{value: -1}   0
{value: NaN}  0

size([0, 0]):
{value: 0}    0
{value: -1}   0
{value: NaN}  0
{value: 1}    0

size([-1, -1]):
{value: 0}    0
{value: -1}   0
{value: NaN}  0
{value: 1}    1

size([-1, 1]):
{value: 0}    0
{value: -1}   0
{value: NaN}  0
{value: 1}    0

size([NaN, NaN]):
{value: 0}    0
{value: -1}   0
{value: NaN}  0
{value: 1}    0

